### PR TITLE
feat: Add support for left and right string functions

### LINKED ISSUE #8934: Support string functions `left` and `right`
# Function: right

The `right` function returns the rightmost `n` characters from the given
`text` string.

``` sql
right(str text, n int) RETURNS text
```

## Parameters

  - `str` - The `text` value from which to extract the rightmost
    characters.

  - `n` - The number of rightmost characters to extract.

## Return Value

The function returns the rightmost `n` characters of the `text` input as
a `text` value.

## Example

To extract the last 3 characters from a given `text` value:

``` sql
SELECT right('hello', 3); -- Output: 'llo'
```

### PR DESCRIPTION
**Pull request by azdxdy**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Add support for `left` and `right` string functions
close #8934

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR contains user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- SQL commands, functions, and operators

### Release note

</details>

### CODE REVIEW

File: src/expr/src/vector_op/string.rs [hunk 1]
```diff
@@ -381,3 +381,87 @@ pub fn quote_ident(s: &str, writer: &mut dyn Write) {
...
+}
+
+/// Returns the last n characters in the string.
+/// If n is a negative value, the function will return an empty string.
+///
+/// # Example
+///
+/// ```slt
+/// query T
+/// select right('RisingWave', 4)
```
Comment (lmatz) [inline src/expr/src/vector_op/string.rs:418]: can we add a test case where the number is larger than the length of string input

**Review by lmatz — APPROVED** [review]
rest LGTM, thanks for your contribution!

File: src/expr/src/vector_op/string.rs (hunk 1 above)
Comment (azdxdy) [reply src/expr/src/vector_op/string.rs:418]: definitely, adding now

Comment (lmatz) [reply src/expr/src/vector_op/string.rs:422]: Thanks!

File: proto/expr.proto [hunk 2]
```diff
@@ -140,6 +140,8 @@ message ExprNode {
     LN = 272;
     LOG10 = 273;
     CBRT = 274;
+    LEFT = 275;
+    RIGHT = 276;
```
Comment (xiangjinwu) [inline proto/expr.proto:144]: nit: move them to 317 and 318

File: src/frontend/src/binder/expr/function.rs [hunk 3]
```diff
@@ -572,6 +572,8 @@ impl Binder {
                 ("radians", raw_call(ExprType::Radians)),
                 ("sqrt", raw_call(ExprType::Sqrt)),
                 ("cbrt", raw_call(ExprType::Cbrt)),
+                ("left", raw_call(ExprType::Left)),
+                ("right", raw_call(ExprType::Right)),
```
Comment (xiangjinwu) [inline src/frontend/src/binder/expr/function.rs:576]: There is a section for string functions below.

File: src/expr/src/vector_op/string.rs [hunk 4]
```diff
@@ -381,3 +381,95 @@ pub fn quote_ident(s: &str, writer: &mut dyn Write) {
     }
     write!(writer, "\"").unwrap();
 }
+
+/// Returns the first n characters in the string.
+/// If n is a negative value, the function will return an empty string.
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/string.rs:386]: Sorry for the misinformation provided in the issue. The PostgreSQL behavior on negative input is actually to return all but last |n| characters.

**Review by xiangjinwu — COMMENTED** [review]
rest lgtm. Thank you for the contribution!

**Comment by xiangjinwu** [comment]
> btw the ci failure on fuzz-test is not related to this PR. It is being fixed and would go away with a 'merge main' after the fix is done. I will let you know when you can 'merge main' to solve it.

Fixed in main.

File: src/expr/src/vector_op/string.rs (hunk 4 above)
Comment (azdxdy) [reply src/expr/src/vector_op/string.rs]: I had question actually why it's not the same as PostgreSQL. Updated the code

**Comment by xiangjinwu** [comment]
Seems the sqlparser is rejecting `left` / `right` as function names so they cannot be used by users end-to-end yet. The docslt tests were actually skipped due to #10749. I will look into the parser error.

**Comment by azdxdy** [comment]
my bad, thank you for fixing this

**Pull request merged**